src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -41,17 +42,27 @@ public object GetService(Type serviceType)
 
         internal object CaptureDisposable(object service)
         {
-            Debug.Assert(!_disposed);
-
             _captureDisposableCallback?.Invoke(service);
 
             if (ReferenceEquals(this, service) || !(service is IDisposable || service is IAsyncDisposable))
             {
                 return service;
             }
 
-            lock (ResolvedServices)
+            lock (_lock)
             {
+                if (_disposed)
+                {
+                    if (service is IDisposable disposable)
+                    {
+                        disposable.Dispose();
+                    }
+                    else
+                    {
+                        ((IAsyncDisposable)service).DisposeAsync().AsTask().GetAwaiter().GetResult();
```

Comment:
The more I think about this and potential downsides, I'm thinking maybe we should fire and forget these...

cc @stephentoub for advice

TL;DR We're trying to take more granular locks in DI so the old code used to wait on what ended up being a global lock per service provider instance. Before disposing during resolve would wait, now we want to remove that global lock from resolve we're changing the behavior to let dispose complete and have any new calls to this method throw an ODE and dispose the thing being created.

Disposing the thing being created means we need to support IAsyncDisposable and IDisposable and this is a synchronous code path. I'm thinking we fire and forget since it's an error case and waiting may cause bad side effects, like a deadlock.

Reply:
The two potential problems with fire-and-forget in a case like this are:
1. The thing being disposed (whatever is calling CaptureDisposable) borrowed some resources that end up still being in use when its Dispose returns, and the caller then rightfully assumes they've regained sole ownership but suffers strange race conditions because they're assumption was wrong.
2. Failures that occur asynchronously end up being masked.

In general (1) is the more impactful issue if it's applicable... is it applicable here?

(2) can be mitigated if there's some logging mechanism that could be employed as part of a continuation.

How common is it for a service to implement IAsyncDisposable but not IDisposable?

Reply:
> How common is it for a service to implement IAsyncDisposable but not IDisposable?

I only know of one thing that does this https://source.dot.net/#Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs,59 😄 

I do think it's extremely rare though. Maybe we just need to avoid the deadlock and call it a day